### ISSUE TITLE
occasional exception when adding/removing views

### ISSUE THREAD

**Issue by jardakotesovec** [paperjs/paper.js#841]
I have use case where I remove/add projects often on the screen, which results in many views add/remove.

On occasions I got exception, because `prevFocus` was pointing to the view that was removed, which leads to failures inside `updateFocus()`. 

So I was trying to understand more logic here
```
	docEvents[mousemove] = function(event) {
		var view = View._focused;
		if (!dragging) {
			// See if we can get the view from the current event target, and
			// handle the mouse move over it.
			var target = getView(event);
			if (target) {
				// Temporarily focus this view without making it sticky, so Key
				// events are handled too during the mouse over.
				// If we switch view, fire one last mousemove in the old view,
				// to give items the change to receive a mouseleave, etc.
				if (view !== target)
					handleMouseMove(view, viewToProject(view, event), event);
				prevFocus = view;
				view = View._focused = tempFocus = target;
			} else if (tempFocus && tempFocus === view) {
				// Clear temporary focus again and update it.
				view = View._focused = prevFocus;
				console.log('Focused view:'+(View._focused && View._focused._id))

				updateFocus();
			}
		}

		if (view) {
			var point = viewToProject(view, event);
			if (dragging || view.getBounds().contains(point))
				tool = handleMouseMove(view, point, event);
		}
	};

```

Is intention that if *view1* has focus and user move with mouse over *view2*, it just temporarily set focus on *view2*, but once mouse gets out of *view2* it got switched back to *view1*?

It seems to me that code is not doing that, because if multiple mouse move events occurs on *view2* `prevFocus` ends up being *view2* and no way to go back to *view1* once mouse leaves *view2*.

Fixing hanging reference to removed view in `prevFocus` can be solved by keeping just view `._id` and retrieve it from `View.viewsById` to be sure of getting existing view. But wanted to check with you first if the actual logic with `prevFocus` is wrong.. or I just don't see it.

Thanks

**Comment by lehni** [comment]
Yes the temporary focus change on mouse activity is intentional. It's only once you click on a view that the focus gets shifted properly. Let me think a little about the issue you're describing here, there might be an easier fix for it.

**Comment by jardakotesovec** [comment]
Thanks for looking into it. I understand this intention, but code does not seems to really do that.

Lets say focus is on view1 and user mouse over on view2.

This branch is called:


```
                prevFocus = view;
                view = View._focused = tempFocus = target;
```

Which is fine, for the first mouse over event, but second one will assign `view2` to `prevFocus`. Because it assigned `target` to `View._focused`. Therefore it can't  go back to `view1`.

**Comment by lehni** [comment]
Yeah I think you're right...

I developed this code ages ago, to deal with this page: http://paperjs.org/features/

The idea was that if you click on a view the view should get the focus properly, but should temporarily borrow it to other views if you hover over them, receiving it back as soon as you leave them.

I think this also links to some other issues people have been reporting, e.g. #686 It's time to look into these issues and solve them.

**Comment by lehni** [comment]
I believe I fixed this in cee3959bfa3baf8837ec06604f1e75f69639be18. Feel free to reopen if it still occurs.
